Allow minting devnet NFTs

benborgers opened the pull request:
This PR adds `NetworkContext` which allows changing between mainnet and devnet for the entire docs site, and persists it to local storage. 

On the Create NFT component itself, there’s a LuxMenu that allows switching network: 

![2022-06-24 at 12 52 22@2x](https://user-images.githubusercontent.com/30215449/175606642-2c4417b1-e5b7-4218-a53e-e95b4b3cdc76.png)

vpontis commented on docs/components/CreateNftSection.tsx:
```diff
@@ -70,126 +72,137 @@ export const CreateNftSection = () => {
     }, 7_500);
   }, [success, setSuccess]);
 
+  const networkContext = useContext(NetworkContext);
```
Let's pull in the function `createUseAppContext` from Luma.

vpontis commented on docs/components/CreateNftSection.tsx:
```diff
@@ -322,6 +342,44 @@ const ImageDropZone = () => {
   );
 };
 
+const NetworkSwitcher = () => {
```
Let's move this to an `atoms` folder.

vpontis commented on docs/components/NetworkContext.tsx:
```diff
@@ -0,0 +1,45 @@
+type Context = {
+  network: Network;
+  setNetwork: (network: Network) => void;
+  networkPrettyName: string;
```
This feels weird. Let's instead have an object `const NETWORK_TO_INFO: Record<Network, {name: string}>`. We usually want to keep state as small as possible.

vpontis commented on docs/components/NetworkContext.tsx:
```diff
@@ -0,0 +1,45 @@
+
+  const setAndPersistNetworkState = (name: Network) => {
+    setNetworkState(name);
+    localStorage.setItem(NETWORK_LOCAL_STORAGE_KEY, name);
```
If we are cacheing state, we should pull that out into a separate hook. Check out `useCookieState` in Luma.

vpontis commented on docs/components/NetworkContext.tsx:
```diff
@@ -0,0 +1,45 @@
+      value={{
+        network: networkState,
+        networkPrettyName:
+          networkState.substring(0, 1).toUpperCase() +
+          networkState.substring(1),
```
So this won't support other networks if we decide to add them in the future. Also, if we are doing something like this, we should either use lodash's capitalize or create a capitalize function. It's nice to read `capitalize` rather than try to figure out exactly what this is doing (which could involve looking up the docs for `substring`.

vpontis reviewed (changes requested):
Let's update the design of the network picker so it's on the top right similar to the "Live Minting Demo" and let's also pull that `<InteractiveWell />` into a component

benborgers commented:
I moved the network switcher and extracted `<InteractiveWell />` (which includes the container, the title “Live Minting Demo”, and the network switcher). 

I switched to using Luma’s `usePersistedState` for interacting with local storage, but I did make a change to the hook d183c7a58bba6148cfa40b810fcc304a761744ab to fix hydration errors. Without that, I was getting errors that the server-side rendered version of the page (which had the default value) didn’t match the client-side version (which had the value from local storage). 

![2022-06-24 at 15 41 52@2x](https://user-images.githubusercontent.com/30215449/175656577-240ad77d-eccf-4664-b2ad-5337ebbf0ce8.png)

vpontis commented on docs/components/InteractiveWell.tsx:
```diff
@@ -0,0 +1,54 @@
+}) => {
+  return (
+    <section className="px-3 pb-3 my-3 rounded">
+      <div className="badge text-xs font-weight-bold">Live Minting Demo</div>
```
We want this to be a reusable component so this shouldn't say `Live Minting Demo`, instead it should have a `{title}` param. 

Let's also not hard code the margin but allow this to accept a `className` prop.

vpontis reviewed (commented):
<img width="712" alt="CleanShot 2022-06-27 at 12 58 35@2x" src="https://user-images.githubusercontent.com/1319079/175995425-3d9c5da0-2e42-44fa-bf6a-a10be11812c7.png">

It feels a bit funny that the title bar isn't blurred out but it also isn't interactive until you sign in. Maybe we can blur it out?

benborgers commented:
Whoops! I had `{title}` but forgot to push it. I added a `className` prop and I’m now blurring everything: 

![2022-06-27 at 13 29 19@2x](https://user-images.githubusercontent.com/30215449/176000463-daeb4b17-a5a7-4808-a6d3-009e002dd608.png)

vpontis commented on docs/components/InteractiveWell.tsx:
```diff
@@ -0,0 +1,135 @@
+      <div className={classNames("children", { blurred })}>{children}</div>
+
+      {!minimal && (
+        <div className={classNames("disconnect", { blurred })}>
```
I'm not a fan of absolute positioning the `Disconnect` buttons since that feels pretty brittle and the Disconnect button isn't part of the chrome of the component, it's part of the inside of the component. Let's move it back to the parent

vpontis commented on docs/pages/_app.tsx:
```diff
@@ -1,6 +1,8 @@
 import { GlowProvider } from "@glow-app/glow-react";
 import "@glow-app/glow-react/dist/styles.css";
 
+import { NetworkProvider } from "../components/NetworkContext";
```
`_app.tsx` has three paths right now. I think it should only have two.

benborgers replied:
Yep! I’m opening a separate PR for that soon, I didn’t want to add more to this PR.
